src/main/java/exception/GlobalExceptionHandler.java

```diff
@@ -0,0 +1,16 @@
+package exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler({NotFoundReservationException.class, InvalidReservationParameterException.class})
+    public ResponseEntity<String> handleIllegalArgumentException(IllegalArgumentException exception) {
```

Comment:
이 핸들러는 정의한 `NotFoundReservationException`,  `InvalidReservationParameterException`에 대한 예외만 처리하는 군요.

step4의 테스트는 위 예외가 아닌 `IllegalArgumentException`가 발생하기 때문에 해당 핸들러(메서드)는 수행 되지 않아요.
> `@ExceptionHandler`에 추가하면 되겠죠?!
\+ 파라미터 역시 타입이 `IllegalArgumentException`이다 보니 어노테이션에서 정의한 예외 역시 처리 되지 않겠어요. 3가지 타입의 예외를 모두 포함하는 추상화된 객체로 선언해보세요!
```suggestion
    @ExceptionHandler({NotFoundReservationException.class, InvalidReservationParameterException.class, IllegalArgumentException.class})
    public ResponseEntity<String> handleIllegalArgumentException(RuntimeException exception) {
```

https://docs.spring.io/spring-framework/reference/web/webmvc/mvc-controller/ann-exceptionhandler.html#mvc-ann-exceptionhandler-args

Reply:
어노테이션에 IllegalArgument를 추가하는 이유는 이해가 되었는데 메서드의 매개변수에 RuntimeException을 추가해야하는 이유는 무엇인가요?

Reply:
Java의 예외 구조 기억나시나요?
<img width="451" alt="Screenshot 2024-11-12 at 11 09 33" src="https://github.com/user-attachments/assets/0b3d0dde-d13e-47b2-8160-8504a82e6a9b">

`Throwable`한 클래스를 구현, 상속하여 커스텀 예외 객체를 선언 할 수 있어요.
그렇게 예외 구조가 생성 됩니다.

지훈님이 예외를 `Global` 하게 처리 하고자 정의한 `@ControllerAdvice`는 앞서 언급드린 "빈"이라는 개념으로 등록 됩니다.
그리고 그 과정에서 `@ControllerAdvice.class`내의 `@ExceptionHandler`라는 어노테이션이 붙은 메서드를 `핸들러`라는 개념으로 등록 해요. (그냥 메서드라고 생각하시죠!)

프로그램 실행 중 웹 요청으로 인한 예외가 발생하면 어떤 `ExceptionHandler`(메서드)가 해당 예외를 처리할 지 찾는 과정이 발생합니다. 이때 예외를 처리하는 메서드, 핸들러를 실행할 때 `InvocableHandlerMethod.class` 의 `getMethodArgumentValues()` 메서드를 거쳐 실행합니다. 그리고 `parameter.getParameterType().isInstance(providedArg)` 라는 로직으로 실행시켜요.
<img width="851" alt="Screenshot 2024-11-12 at 11 31 05" src="https://github.com/user-attachments/assets/c6a97a47-97da-4fb5-a7e3-d5b336c57111">

**그래서 `handler`의 파라미터가 처리 해야하는 예외 타입과  상속/구현이(`isInstance()`) 되어있어야 합니다.**

다시 예외 구조로 돌아와서, `IllegalArgumentException.class`는 `IllegalArgumentException`에 해당하는 예외 타입만 찾겠죠.
어노테이션으로 선언한, 즉 핸들러로 등록하여 처리 하기로 정의 한 예외 타입들이 모두 상속/구현하고 있는 예외를 파라미터로 정의해야 실행 가능해요.

조금 내용이라 추가적인 설명이 필요하다면 언제든지 알려주세요! 텍스트로 전달하는데 한계가 있어 디스코드 라이브로도 요청 주셔도 됩니다!

Reply:
그래서 handler의 파라미터가 처리 해야하는 예외 타입과 상속/구현이(isInstance()) 되어있어야 합니다.
-> 이 부분에서 이해가 되었습니다! 너무 생소한 내용이라 이 부분은 미션 진행 이후에도 프로젝트를 하면서 익숙해질 필요가 있을 것 같아요!